src/rulesets/Core/core.jl

```diff
@@ -32,3 +32,34 @@ function rrule(::typeof(ifelse), c, a::Number, b::Number)
     ifelse_pullback(Δ) = (NoTangent(), NoTangent(), ifelse(c, Δ, zero(Δ)), ifelse(c, zero(Δ), Δ))
     return ifelse(c, a, b), ifelse_pullback
 end
+
+function rrule(
+    ::typeof(Core._apply_iterate), ::typeof(iterate), f::F, args...) where F
+    # flatten nested arguments
+    flat = []
+    for a in args
+        push!(flat, a...)
+    end
+    # apply rrule of the function on the flat arguments
+    y, pb = rrule(f, flat...)
```

Comment:
This needs to use `rrule_via_ad`
Since nothing promised that an `rrule` will exist.

This mean this rule needs a config

Reply:
Does this have overhead?

I'd worry a little bit that providing a rule for some deep-internal function like this would insert ChainRules into lots of places where it needn't go. And I wonder whether this shouldn't be handled by the AD itself? It's about how Julia structures function calls, rather than some mathematical property.

Reply:
Yeah, `Core._apply_iterate()` may be deep-internal, but as far as I understand it arises in user-level code every time we use the splatting operator. If a user writes:

```julia
reshape(x, sz...)
```
they expect this call to be handled by AD independently of what it looks like in the lowered form. And in our case it looks looks like `Core._apply_iterate(iterate, reshape, sz)`.

I thought about replacing `Core._apply_iterate(iterate, f, args...)` directly with `f(f, a, b, c)` before running the reverse pass in the AD, but in general case it's not possible to rewrite array of tuples `args` into flat argument `a, b, c` because you may not have access to details of the tuples. So it's unclear how to represent calls to `_apply_iterate()` in a more ChainRules-friendly way. 

One way to stop this `rrule` from getting to the "wrong places" is to restrict it only to specific functions `f`, e.g.:

```julia
function rrule(
    ::typeof(Core._apply_iterate), ::typeof(iterate), f::typeof(reshape), x, sz...)
   ...
end
```
I suppose it will be safer and maybe even more efficient in some cases. On the downside, if we introduce, say, 20 such rules and in some future versions Julia changes the way splatting is handled, we will need to update all the 20 rules for that version. But perhaps we can talk about it when we have 20 such rules :smile:

Reply:
The more I think about handling it on the AD side, the more I like this idea. In particular, in Yota I have an alternative rule engine ("d-rules") that are checked before ChainRules. Since so far I'm the only person to complain about missing rrule for `_apply_iterate()`, it makes sense to use that machinery instead. If it works well, we can postpone the more generic solution until we get more use cases.